Configurable get-words command, python3.4 compatibility, bring back async

balta2ar opened the pull request:
Hello! I thought about breaking this into smaller pull requests, but the code is rather small, so I'm not sure it was worth it. This patch does the following:

1. Adds a deoplete option `g:deoplete#sources#webcomplete#script` to configure the command that prints a list of words (instead of default `sh/webcomplete`)

2. Replaces subprocess.run with subprocess.check_output because there is no `run` yet in Python3.4.

3. Brings back the async but in a rather different way. I have my own script that prints a list of words from the active pages of my browsers, but unfortunately it's not very snappy. It executes in about 1..2 sec, so it's a bit annoying in vim when I have to wait for the words to show up. Instead, I cache the results and immediately return the cache, and also send "retrieve the words again" signal to the background thread. Even though technically we potentially receive results from the previous query, in practice it's not a problem because 1. active page does not change that often 2. you are likely to trigger another "retrieve the words again" signal when you start a new word and `_is_same_context` returns False.

balta2ar committed:
Disable async, add logging and fake data for reproducability

balta2ar committed:
Fetch words in background thread

balta2ar committed:
Use threading.Event instead of Queue

balta2ar committed:
Add deoplete var deoplete#sources#webcomplete#script

balta2ar committed:
Cleanup logging and empty line

balta2ar committed:
add basic integration with ncm2

thalesmello commented on pythonx/ncm2_webcomplete.py:
```diff
@@ -0,0 +1,57 @@
+        # filedir = dirname(abspath(__file__))
+        # projectdir = abspath(join(filedir, pardir))
+        # join(projectdir, 'sh', 'webcomplete')
+        self.__script = expanduser(expandvars('~/.local/bin/bt words'))
```
I noticed you have hardcoded your script in the NCM2 source you've implemented, which makes it unusable outside your computer.

Is there a reason why you didn't get the variable from the var configuration, as you did in the deoplete source file?

balta2ar committed:
Read script from g:ncm2_webcomplete_script var instead of hard-coding

balta2ar committed:
Remove look from source dict name

balta2ar commented:
@thalesmello Thanks for reviewing my pull request! I totally forgot about it and just switched to my forked repo. Initially I was using deoplete but then switched to ncm2, and added a quick and dirty solution for ncm2 source in my repo. Anyway, I've cleared up the mess and introduced a variable to configure the script path in case you would like to merge it.

thalesmello commented on pythonx/ncm2_webcomplete.py:
```diff
@@ -0,0 +1,52 @@
+        filedir = dirname(abspath(__file__))
+        projectdir = abspath(join(filedir, pardir))
+        self.__script = join(projectdir, 'sh', 'webcomplete')
+        self.__script = self.nvim.eval('g:ncm2_webcomplete_script') \
```
This line of code throws an error.

Even though you have a fallback for the case the variable `g:ncm2_webcomplete_script`, it still throws an error when it doesn't exist.

```
[ncm2_webcomplete@yarp] Traceback (most recent call last):
[ncm2_webcomplete@yarp]   File "/Users/thalesmello/.config/nvim/plugged/nvim-yarp/pythonx/yarp.py", line 59, in <module>
[ncm2_webcomplete@yarp]     module_obj = importlib.import_module(module)
[ncm2_webcomplete@yarp]   File "/usr/local/Cellar/python/3.7.0/Frameworks/Python.framework/Versions/3.7/lib/python3.7/importlib/__init__.py", line 127, in import_module
[ncm2_webcomplete@yarp]     return _bootstrap._gcd_import(name[level:], package, level)
[ncm2_webcomplete@yarp]   File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
[ncm2_webcomplete@yarp]   File "<frozen importlib._bootstrap>", line 983, in _find_and_load
[ncm2_webcomplete@yarp]   File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
[ncm2_webcomplete@yarp]   File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
[ncm2_webcomplete@yarp]   File "<frozen importlib._bootstrap_external>", line 728, in exec_module
[ncm2_webcomplete@yarp]   File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
[ncm2_webcomplete@yarp]   File "/Users/thalesmello/.config/nvim/plugged/webcomplete.vim/pythonx/ncm2_webcomplete.py", line 49, in <module>
[ncm2_webcomplete@yarp]     source = Source(vim)
[ncm2_webcomplete@yarp]   File "/Users/thalesmello/.config/nvim/plugged/webcomplete.vim/pythonx/ncm2_webcomplete.py", line 30, in __init__
[ncm2_webcomplete@yarp]     self.__script = self.nvim.eval('g:ncm2_webcomplete_script') \
[ncm2_webcomplete@yarp]   File "/usr/local/lib/python3.7/site-packages/neovim/api/nvim.py", line 280, in eval
[ncm2_webcomplete@yarp]     return self.request('nvim_eval', string, **kwargs)
[ncm2_webcomplete@yarp]   File "/usr/local/lib/python3.7/site-packages/neovim/api/nvim.py", line 170, in request
[ncm2_webcomplete@yarp]     res = self._session.request(name, *args, **kwargs)
[ncm2_webcomplete@yarp]   File "/usr/local/lib/python3.7/site-packages/neovim/msgpack_rpc/session.py", line 100, in request
[ncm2_webcomplete@yarp]     raise self.error_wrapper(err)
[ncm2_webcomplete@yarp] neovim.api.nvim.NvimError: b'Vim:E121: Undefined variable: g:ncm2_webcomplete_script'
[ncm2_webcomplete@yarp] Job is dead. cmd=['python3', '-u', '/Users/thalesmello/.config/nvim/plugged/nvim-yarp/pythonx/yarp.py', '/var/folders/4f/4r6phsjj0jn19qywlx5zxrjc0000gn/T/nvimDey8pj/0', 8, 'ncm2_webcomplete']
```

An alternative might be `self.__script = self.self.nvim.eval("get(g:,'ncm2_webcomplete_script', '')" or self.__script`

balta2ar committed:
Make sure default value for g:ncm2_webcomplete_script is set

balta2ar commented:
@thalesmello Defined in autoload/ncm2_webcomplete.vim instead so that all variables are in one place.

thalesmello commented:
Thanks for your contribution @balta2ar
